### ISSUE TITLE
Inconsistent results between marginal_means() and emmeans() for glm() with a spline variable

### ISSUE THREAD

**Issue by chiungming** [vincentarelbundock/marginaleffects#831]
I have noticed that marginal_means() and emmeans() produce different results when a spline variable is included in glm(). The results from both functions are in agreement for lm(). 

```
library(splines)
library(marginaleffects)
library(emmeans)

mod <- glm(mpg ~ bs(hp) + factor(cyl), data = mtcars, family = "gaussian")

emmeans(mod, "cyl")
# cyl emmean   SE df lower.CL upper.CL
# 4   21.6 2.10 26     17.3     25.9
# 6   18.1 1.31 26     15.4     20.8
# 8   16.6 1.45 26     13.6     19.6
# 
# Confidence level used: 0.95 

predictions(mod, newdata = datagrid(cyl = 4))
# Estimate Pr(>|z|)    S 2.5 % 97.5 %  hp cyl
# 21.6   <0.001 79.8  17.5   25.7 147   4
# 
# Columns: rowid, estimate, p.value, s.value, conf.low, conf.high, mpg, hp, cyl 

predictions(mod, newdata = datagrid(cyl = 6))
# Estimate Pr(>|z|)     S 2.5 % 97.5 %  hp cyl
# 18.1   <0.001 141.4  15.5   20.7 147   6
# 
# Columns: rowid, estimate, p.value, s.value, conf.low, conf.high, mpg, hp, cyl 

predictions(mod, newdata = datagrid(cyl = 8))
# Estimate Pr(>|z|)    S 2.5 % 97.5 %  hp cyl
# 16.6   <0.001 98.3  13.8   19.5 147   8
# 
# Columns: rowid, estimate, p.value, s.value, conf.low, conf.high, mpg, hp, cyl 


marginal_means(mod, variables = "cyl") 
# Term Value Mean Pr(>|z|)     S 2.5 % 97.5 %
# cyl     6 26.6   <0.001  60.8  20.8   32.5
# cyl     4 30.1   <0.001 214.8  26.7   33.6
# cyl     8 25.2   <0.001  34.2  17.6   32.7
# 
# Results averaged over levels of: cyl 
# Columns: rowid, term, value, cyl, estimate, p.value, s.value, conf.low, conf.high, hp, mpg, wts 

predictions(mod, newdata = datagrid(cyl = unique))
# Estimate Pr(>|z|)     S 2.5 % 97.5 %  hp cyl
# 26.6   <0.001  60.8  20.8   32.5 147   6
# 30.1   <0.001 214.8  26.7   33.6 147   4
# 25.2   <0.001  34.2  17.6   32.7 147   8
# 
# Columns: rowid, estimate, p.value, s.value, conf.low, conf.high, mpg, hp, cyl 



mod <- lm(mpg ~ bs(hp) + factor(cyl), data = mtcars)
marginal_means(mod, variables = "cyl") 
# Term Value Mean Std. Error    z Pr(>|z|)     S 2.5 % 97.5 %
# cyl     6 18.1       1.31 13.8   <0.001 141.4  15.5   20.7
# cyl     4 21.6       2.10 10.3   <0.001  79.8  17.5   25.7
# cyl     8 16.6       1.45 11.4   <0.001  98.3  13.8   19.5
# 
# Results averaged over levels of: cyl 
# Columns: rowid, term, value, cyl, estimate, std.error, statistic, p.value, s.value, conf.low, conf.high, hp, mpg, wts 

predictions(mod, newdata = datagrid(cyl = unique))
# Estimate Std. Error    z Pr(>|z|)     S 2.5 % 97.5 %  hp cyl
# 18.1       1.31 13.8   <0.001 141.4  15.5   20.7 147   6
# 21.6       2.10 10.3   <0.001  79.8  17.5   25.7 147   4
# 16.6       1.45 11.4   <0.001  98.3  13.8   19.5 147   8
# 
# Columns: rowid, estimate, std.error, statistic, p.value, s.value, conf.low, conf.high, mpg, hp, cyl 
```

**Comment by vincentarelbundock** [comment]
Thanks a lot for the report. This is very weird (especially the examples below). I’ll look into it, but have a busy 2-3 weeks ahead, so I may not be able to diagnose for a little bit.

``` r
library(splines)
library(marginaleffects)

mod <- glm(mpg ~ bs(hp) + factor(cyl), data = mtcars, family = "gaussian")

predictions(mod, newdata = datagrid(cyl = c(4, 6, 8)))
# 
#  Estimate Pr(>|z|)     S 2.5 % 97.5 %  hp cyl
#      30.1   <0.001 214.8  26.7   33.6 147   4
#      26.6   <0.001  60.8  20.8   32.5 147   6
#      25.2   <0.001  34.2  17.6   32.7 147   8
# 
# Columns: rowid, estimate, p.value, s.value, conf.low, conf.high, mpg, hp, cyl

predictions(mod, newdata = datagrid(cyl = 8))
# 
#  Estimate Pr(>|z|)    S 2.5 % 97.5 %  hp cyl
#      16.6   <0.001 98.3  13.8   19.5 147   8
# 
# Columns: rowid, estimate, p.value, s.value, conf.low, conf.high, mpg, hp, cyl
```

The discrepancy only arises when using `bs()`.

**Comment by vincentarelbundock** [comment]
Thanks again for the report.

The problem was that `stats::model.matrix(model,newdata=df)` returns an all-zero matrix when `bs()` is used in the formula. I coded a workaround which should be slightly slower but produce the same results as `emmeans`. Available now on Github in version 0.13.0.9012. Will try to release to CRAN in the next couple weeks.

``` r
library(splines)
library(marginaleffects)
library(emmeans)

mod <- glm(mpg ~ bs(hp) + factor(cyl), data = mtcars, family = "gaussian")

emmeans(mod, "cyl")
#  cyl emmean   SE df lower.CL upper.CL
#    4   21.6 2.10 26     17.3     25.9
#    6   18.1 1.31 26     15.4     20.8
#    8   16.6 1.45 26     13.6     19.6
# 
# Confidence level used: 0.95

marginal_means(mod, variables = "cyl")
# 
#  Term Value Mean Pr(>|z|)     S 2.5 % 97.5 %
#   cyl     6 18.1   <0.001 141.4  15.5   20.7
#   cyl     4 21.6   <0.001  79.8  17.5   25.7
#   cyl     8 16.6   <0.001  98.3  13.8   19.5
# 
# Results averaged over levels of: cyl 
# Columns: rowid, term, value, cyl, estimate, p.value, s.value, conf.low, conf.high, hp, mpg, wts

predictions(mod, newdata = datagrid(cyl = unique))
# 
#  Estimate Pr(>|z|)     S 2.5 % 97.5 %  hp cyl
#      18.1   <0.001 141.4  15.5   20.7 147   6
#      21.6   <0.001  79.8  17.5   25.7 147   4
#      16.6   <0.001  98.3  13.8   19.5 147   8
# 
# Columns: rowid, estimate, p.value, s.value, conf.low, conf.high, mpg, hp, cyl
```

**Comment by chiungming** [comment]
Thank you for your great work!
